docs/src/writing-constructs.rst

```diff
@@ -0,0 +1,169 @@
+  reasonable value if not supplied. This may involve creating other resources
+  as part of constructing this construct. For example, all resources that
+  require a role allow passing in a **Role** object, but if the user does
+  not supply one an appropriate **Role** object is constructed in-place.
+* Use contextual defaulting between properties; the value of one property may
+  affect sensible defaults for other properties. *Example*: enableDnsHostnames
+  and and enableDnsSupport.  dnsHostnames requires dnsSupport, only yell if the
+  user has *explicitly* disabled dnsSupport, but tried to enable dnsHostnames.
+  A user expects things to just work - make that the case unless they tell us
+  otherwise.
+* Make the user thinkg about concepts, not implementation detail; for example,
+  if establishing an association between two resources (such as a **Topic**
+  and a **Queue**) requires multiple steps (in this case, creating a
+  **Subscription** but also setting appropriate IAM permissions), make
+  both things happen in a single call (to **subscribeQueue()**). The user
+  should be mentally thinking about the result they're trying to achieve,
+  not the various steps necessary to get there.
+* Make sure you are not introducing new concepts or terminology. For example
+  don't be temped to replace SQS's **dataKeyReusePeriod** with **keyRotation**
+  because it will be hard for people to diagnose problems. They won't be able to
+  just search for *sqs dataKeyReuse* and find topics on it.
+* Indicate units of measurement in property names that don't use a strong type.
+  Use **milli**, **sec**, **min**, **hr**, **Bytes**, **KiB**, **MiB**, **GiB**,
+  etc.
+* Optimize for the common case; example: **Fleet** accepts a **VPC** and deploys
+  in the private subnet by default because that's the common case, but has an
+  option to **placementOptions** for the special case.
+* Use Typescript; in order for construct libraries to be reusable across
+  programming languages (using jsii), they need to be authored in a language
+  that can compile to a jsii assembly. At the moment, the only supported
+  language is TypeScript.
+
+Implementation Details
+======================
+
+* Every construct consist of an exported class (**MyConstruct**) and an exported
+  interface (**MyConstructProps**) that defines the parameters for these
+  classes.
+* Constructs are (mostly) immutable; sometimes we use mutating methods if it
+  makes sense or reads better (for example, for event handlers we will write
+  `builder.onBuild(topic)`), but configuration props are usually passed into
+  the constructor, after which they cannot be changed anymore.
+* Most of the logic happens in the constructor; the constructor will build up
+  the state of the construct (what children it has, which ones are always
+  there and which ones are optional, etc.).
+* Validate as early as possible; throw an `Error` in the constructor if the
```

Comment:
I think it's one place we could specify the ideal failure modes:
1. Best: Wrong can't compile, because type-safety guarantees correctness
2. Good: Runtime check everything the type checker can't enforce and fail early
3. Okay: Synth-time validate everything that can't be checked at construction time
4. Bad (horrible?): Fail in CloudFormation